### ISSUE TITLE
Attachment is not stored 

### ISSUE THREAD

**Issue by matmue** [RealRaven2000/FiltaQuilla#93]
Hello

OS: MacOSX 10.13.6
Thunderbird  68.12.1
FiltaQuilla      2.1.2

**Problem:** 
I'm not able to save or splitt-off an email attachment into a predefined folder with FiltaQuilla.

I activate the save, split off, print and execute options in then FiltaQuilla preferences and restart Thunderbird.
I test my filter workflow manually and the print and exec options work, but the save or split-off option not. 

**Background:**
I want to automatically save / splitt-off email attachments in sender-dependent folders with an email filter.
My desired workflow:
1. Filter identify sender email ( success )
2. Mark Email ( success )
3. Handle Email-attachment 
    a) save in a sender-dependend predefined file-folder                                         ( using FiltaQuilla "save": did not work )
    b) add date and sender to the filenname, ex. 2021_02_06-Google-Invoice.pdf (using FiltaQuilla "exec": start program)
    c) print Email                                                                                                          ( using FiltaQuilla "print": work )
    c) print Attachment                                                                                                (depend on "save")                                                   
4. Move Email to sender-dependend email-folder ( success )

Because FiltaQuilla can't change the filename of the attachment I use the "excecute program" option to start a shell script. But of course I need the attachment as a file first.

Thanks for all hints
Matthias

**Comment by RealRaven2000** [comment]
From what I have heard so far the attachments cannot be used with move to folder, can you try replacing that with "move later"?

**Comment by matmue** [comment]
Thanks for the quick response!

Meanwhile I reduce the filter for test purpose to only one action: "save". 
Unfortunately nothing is saved

**Comment by RealRaven2000** [comment]
> Thanks for the quick response!
> 
> Meanwhile I reduce the filter for test purpose to only one action: "save".
> Unfortunately nothing is saved

- Does it work when you execute the filter manually?
- Make sure to run the filter "after junk controls"
- If on IMAP, make sure the whole emails are downloaded.

**Comment by matmue** [comment]
I test the filter manually.
There is a prepared Email in the Inbox and I start the filter manually.
It doesn't work.

**Comment by RealRaven2000** [comment]
check in the filter logs whether it was running at least. you can also show
JavaScript error  console during filter execution to catch a log there in
case exceptions are thrown

**Comment by matmue** [comment]
I enable the Debug option in FiltaQuilla, restart Thunderbird and use this manual started filter rule:
1. save attachment to file-folder 
2. start shell-program in file-folder 
3. copy email to email-folder

The file-folder /Users/mmuench/Daten/D_ib-muench/Berichtsjahr_2022 is writable for me and I can manual save attached email-attachments there. The result of a manual filter start is:
1. no saved attachment in file-folder
2. shell program start and end successfull
3. email are copied to email-folder

Therefore the filter starts and run. 

The log of the Thunderbird Filter is 
![2021_02_08-Thunderbird-Filterlog](https://user-images.githubusercontent.com/22960953/107218233-a332c800-6a0f-11eb-98fc-ddd64330ecf0.png)


The log of the Developer / Thunderbird Error log console is
![2021_02_08-ThunderbirdErrorLog](https://user-images.githubusercontent.com/22960953/107218255-ad54c680-6a0f-11eb-9dae-1185e66d0da8.png)

It seems to me, FiltaQuilla start a "save"-call but didn't save the file in the given file-folder. I search for other places, but I can't find the attached file in any other folders at the hdd.

**Comment by RealRaven2000** [comment]
> The log of the Developer / Thunderbird Error log console is
> ![2021_02_08-ThunderbirdErrorLog](https://user-images.githubusercontent.com/22960953/107218255-ad54c680-6a0f-11eb-9dae-1185e66d0da8.png)
> 
> It seems to me, FiltaQuilla start a "save"-call but didn't save the file in the given file-folder. I search for other places, but I can't find the attached file in any other folders at the hdd.

```
try {
  let ds = aMsgHdr.date / 1000,
      msgDate = new Date(ds),  // this is cast to string for some stupid reason, so it's not useful.
      msgSubject = aMsgHdr.subject;
  if (util.isDebug) {
    util.logDebug('saveAttachmentCallback_callback');
    debugger;
  }
  // note: for some reason I could not use msgDate as it is treated here as a string not a Date object...
  // the only workaround was to create new date objects at each step and call its functions directly:
  let nicedate = " " + (new Date(ds)).getFullYear() + "-" + ((new Date(ds)).getMonth()+1) + "-" + (new Date(ds)).getDate()  + " " +  (new Date(ds)).getHours() + ":" + (new Date(ds)).getMinutes();
  if (!this.detach) {
    for (let j = 0; j < this.attachments.length; j++) {
      try {
        let attachment = this.attachments[j];
        if (attachment.url.startsWith("file:")) {
          util.logToConsole("Attachment for '" + msgSubject + "' from " 
            + nicedate + " was already removed from mail - last seen at this location:\n" 
            + attachment.url);
          continue;
        }
        // create a unique file for this attachment
        let uniqueFile = this.directory.clone();
        uniqueFile.append(attachment.name);
        let txt = "Save attachment [" + j + "] to " + uniqueFile.path +
            "...\n msgURI=" + this.msgURI +
            "\n att.url=" + attachment.url +
            "\n att.ncontentType=" + attachment.contentType;
        util.logDebug(txt);
        txtStackedDump += txtStackedDump + txt + "\n";
        uniqueFile.createUnique(Ci.nsIFile.NORMAL_FILE_TYPE, 0o600);
        messenger.saveAttachmentToFile(uniqueFile, attachment.url, this.msgURI,
                                       attachment.contentType, null);
      }
      catch (ex) {
        util.logException("SaveAttachmentCallback\n" + txtStackedDump, ex);
      }
    }
  }
  else
  {
    // ...  (detach case)
  }
}
catch (ex) {
  util.logException("SaveAttachmentCallback\n" + txtStackedDump, ex);
}

```
    
Indeed, I am missing the next log step
Save attachment [0] to _filepath_ ...
but I think that's only in the Tb78 version... :(

**Comment by RealRaven2000** [comment]
Here is a quick version for tB68 with the extra logging for attachments added:

[filtaquilla-2.2pre2.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5944131/filtaquilla-2.2pre2.zip)
to install, drag the zip file into Tb ADd-ons Manager.

**Comment by matmue** [comment]
After installing, restart Thunderbird, start filter here the console log.

![2021_02_08-ThunderbirdErrorLog2](https://user-images.githubusercontent.com/22960953/107233992-78eb0580-6a23-11eb-93b6-c73bccd4604a.png)

Seems to be identical messages, or?

**Comment by RealRaven2000** [comment]
> After installing, restart Thunderbird, start filter here the console log
> 
> Seems to be identical messages, or?

Looks like there were no attachments - if the length of the `mimeMessage.allAttachments` array is 0 you would see no further messages. Let's try this one to confirm:

[filtaquilla-2.2pre3.zip](https://github.com/RealRaven2000/FiltaQuilla/files/5944674/filtaquilla-2.2pre3.zip)

for some reason the actions don't actually show up on my Thunderbird 68 system at all :(
And unfortunately my Tb68 test profile has already be upgraded to 78.

**Comment by RealRaven2000** [comment]
I did a test run on my production pile (with debug mode enabled on the "About FiltaQuilla/Help" page of settings ) here is the output with 3 attachments:

```
FiltaQuilla 15:52:8.603  [25380 ms]   
saveAttachment() - target directory exists:
E:\Dev\Mozilla\Dev\FiltaQuilla\test\att
quickFilters [logTime init]
popupAlert(Apply filters to selected messages…, quickFilters)
FiltaQuilla 15:52:8.631  [28 ms]   
3
FiltaQuilla 15:52:8.631  [0 ms]   
Save attachment [0] to E:\Dev\Mozilla\Dev\FiltaQuilla\test\att\license.txt...
 msgURI=mailbox-message://axelg@pop.vodafone.ie/Development/Mozilla/Addons/filtaQuilla#827
 att.url=mailbox:///W:/_Tb%20Profiles/12343uiyg.default/Mail/pop.vodafone.ie/Development.sbd/Mozilla.sbd/Addons.sbd/filtaQuilla?number=827&part=1.2&filename=license.txt
 att.ncontentType=text/plain
FiltaQuilla 15:52:8.940  [309 ms]   
Save attachment [1] to E:\Dev\Mozilla\Dev\FiltaQuilla\test\att\readme.txt...
 msgURI=mailbox-message://axelg@pop.vodafone.ie/Development/Mozilla/Addons/filtaQuilla#827
 att.url=mailbox:///W:/_Tb%20Profiles/12343uiyg.default/Mail/pop.vodafone.ie/Development.sbd/Mozilla.sbd/Addons.sbd/filtaQuilla?number=827&part=1.3&filename=readme.txt
 att.ncontentType=text/plain
FiltaQuilla 15:52:8.941  [1 ms]   
Save attachment [2] to E:\Dev\Mozilla\Dev\FiltaQuilla\test\att\build.manifest...
 msgURI=mailbox-message://axelg@pop.vodafone.ie/Development/Mozilla/Addons/filtaQuilla#827
 att.url=mailbox:///W:/_Tb%20Profiles/12343uiyg.default/Mail/pop.vodafone.ie/Development.sbd/Mozilla.sbd/Addons.sbd/filtaQuilla?number=827&part=1.4&filename=build.manifest
 att.ncontentType=text/plain
```

these 3 files were saved correctly.

**Comment by matmue** [comment]
I install 2.2pre4 an test it without success. 
I create a new email with another pdf attachment, even without success.

No attachment is saved but the error log changed.

![2021_02_08-ThunderbirdErrorLog3](https://user-images.githubusercontent.com/22960953/107283107-3a257180-6a5c-11eb-91ce-8074127b00d0.png)

**Comment by RealRaven2000** [comment]
> No attachment is saved but the error log changed.
> 
> ![2021_02_08-ThunderbirdErrorLog3](https://user-images.githubusercontent.com/22960953/107283107-3a257180-6a5c-11eb-91ce-8074127b00d0.png)

So this happens internally in Thunderbird core code - which means I cannot add any error message. It happens in this function:
https://searchfox.org/comm-esr68/source/mailnews/db/gloda/modules/mimemsg.js#179

the only way to see what is wrong would be to debug the Thunderbird function `MsgHdrToMimeMessage()` on the PC where it happens. So the function throws (it shouldn't) and never gets to call the callback function that is in charge of saving the file. Not sure what it can be caused by, my guesses are
1) broken message
2) mail not fully downloaded
3) something encrypted?
4) configuration problem on IMAP.

tbh I have no idea. I only know that attachments don't work well when called by the filtering mechanism

**Comment by RealRaven2000** [comment]
To find out what it happening on your machine - you could debug this yourself using the Developer Toolbox:

- Tools / Developer Tools / Developer Toolbox (or hit Ctrl+Shift+I)
- Click the "Debugger" tab
- Ctrl+P to select a file
- enter filtaquilla in the search box:
  ![image](https://user-images.githubusercontent.com/6314674/107350753-b900c580-6ac1-11eb-9acf-a22c251d8fa3.png)
- Set a breakpoint at line 626 (the call to `MsgHdrToMimeMessage()`) by clicking on the line number:
  ![image](https://user-images.githubusercontent.com/6314674/107350894-e483b000-6ac1-11eb-8ca0-0cfe3e2c88fe.png)
- execute the filter - once the debugger lands at this line, trace into the function `MsgHdrToMimeMessage()` using the button on the top right:
  ![image](https://user-images.githubusercontent.com/6314674/107351068-1432b800-6ac2-11eb-958e-4157d3b30247.png)
- you will now be inside the Thunderbird module and can single step through the code:
![image](https://user-images.githubusercontent.com/6314674/107351144-2d3b6900-6ac2-11eb-97bd-8dc7ab271c8d.png)
- take a note of the Thunderbird line where the exception happens. (it will jump out of the code and end in the next `catch { }` block where exceptions are handled

**Comment by matmue** [comment]
Thank you very much for the hint, I will try that.

**Comment by shikousa** [comment]
I'm having the same issue. The attachment of a message with a certain header shall automatically be stored in a local folder. This does not working on incoming messages, but I get a log entry in the protocoll. If I press run now over all messages then it works.

**Comment by RealRaven2000** [comment]
I think everyone on this thread "only"  experiences this problem when Thunderbird starts up, so I am marking as duplicate of #91 and closing this one.

(I have 17 issues related to attachments, it's a total mess. And it's very likely bad implementation on Thunderbird's side.)
